### ISSUE TITLE
Amplify Api not working properly when using filter on ModelQueries.list() as well as on custom Graphql Query

### ISSUE THREAD

**Issue by uzairleo** [aws-amplify/amplify-flutter#3873]
### Description

Hi there, we are using amplify Api backed by dynamodb for our use-case and there is a scenario where we implement search on the user table and fetch records on the basis of user input email, which is based on Amplify Api Query where we actually apply some filter on that Query using the `where` field of ModelQueries.lis() method 
  ```
 final request = ModelQueries.list(model, where: where, limit: limit);
      final response = await Amplify.API.query(request: request).response;
```
It was working awesome when the records were small But now it's behaving weirdly as it is not fetching and even not showing an empty response on some of the records that are available in dynamodb and one can pull through dynamodb but the Query fails to fetch and even show at least empty response.

After looking into some of the amplify-js threads on the same topic on GitHub issues I found that using custom Query is something emphasized for our use-case such as search or advanced search implementation. so I replaced ModelQueries.list() based implementation into Custom Graphql Query and that resolved the issue for us, this custom query was able to return every user which matched with user-provided email/literals. 
 ```
query SearchUsers(\$email: String!, \$id: ID!,\$isCompleted: Boolean, \$nextToken: String) {
          listTotemUSERS(
            filter: {
               email: { contains: \$email }, 
               id: { ne: \$id },
               isCompleted: { eq: \$isCompleted }
               },
             limit: 9999, nextToken: \$nextToken ) {
            items {
              id
              email
              name
              pmtRefNo
              phoneNumber
              isCompleted
            }
             nextToken
          }
        } 
```
But this fix brings another issue as the fix works awesome in the dev/staging env of Amplify where we have a small number of records, not bulks.  But when we tried to migrate the implementation to prod env of amplify for testing purposes where we have a bulk of records then we found that the Custom Graphql Query is also not fully working as its not fetching the record when given the full email But it does fetch record when providing first one to 3ree literals of user email Also keeping in mind this behavior is only limited now for all those user records that were recently not working not even returning an empty response with ModelQueries.list() . 
Another thing that causes confusion is the same records that were not pulled by the above methods were also not pulled through the amplify content manager test Graphql console by providing either the same query or a different query of List it can only be fetched through dynamodb and with LIST  query when providing email partially. Why is it behaving like that?

### Categories

- [ ] Analytics
- [X] API (REST)
- [X] API (GraphQL)
- [ ] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Notifications (Push)
- [ ] Storage

### Platforms

- [X] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.3.10

### Amplify Flutter Version

0.6.10

### Deployment Method

Amplify CLI

### Schema

```GraphQL
type User @model @auth(rules: [{allow: public}]) {
  id: ID!
  name: String!
  email: String
  declined: Boolean
}
```

**Comment by uzairleo** [comment]
@Jordan-Nelson @jacoblogan

**Comment by uzairleo** [comment]
@dnys1

**Comment by uzairleo** [comment]
@khatruong2009 any updates Guys ?

**Comment by Equartey** [comment]
Hey @uzairleo, sorry you're experiencing this issue.

To summarize the issue, you are querying a list of `User`s and want to filter the results by passing a full email string. However, you're not receiving any matches for the full string, but you are receiving results when you pass a partial email string. This is working on a small data set, but not a large one. Is all this correct? 

If so, you're implementation is almost correct. To have efficient queries on large data sets via AppSync and DynamoDB we recommend you setup global secondary indexes. Can you review this [comment](https://github.com/aws-amplify/amplify-flutter/issues/2006#issuecomment-1210728199). I'll also call out you'll need to adjust the limit parameter so you are not capping your scans at 9999.

Please don't hesitate to reach out if you have more questions.

**Comment by uzairleo** [comment]
yeap you are right @Equartey Thanks for responding. Can you please share the official setup of up GSI doc for Flutter , Also I do have one Question regarding the implementation of this GSI. 

1-If we go with setting up GSI I believe we have to modify Graphql schema if that the case then I have a concern about this fix that is The fix will only work for new users but for an old user it won't work because that's already registered with old schema without GSI let me know if I am wrong @Equartey

**Comment by Equartey** [comment]
Hi @uzairleo 

Setting up GSI in Flutter has two main steps, CLI/Schema setup and the Flutter implementation. 

1. For the required schema updates please reference this [doc](https://docs.amplify.aws/cli/graphql/data-modeling/#configure-a-secondary-index).

2. For the Flutter implementation, we don't have an explicit guide for filtering use case, but this [sorting guide](https://docs.amplify.aws/guides/api-graphql/query-with-sorting/q/platform/flutter/) is pretty close. Let me know if you have further questions.

As for updates to the schema, DynamoDB handles backfill of the new index ([reference](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/GSI.OnlineOps.html#GSI.OnlineOps.Creating)). However, your schema has `email` marked as an optional property. Querying on GSI won't work on `User` objects that did not provide an email. This is called a [sparse index](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-indexes-general-sparse-indexes.html) and is still valuable feature.

In short, existing users should be able to use the new GSI once setup correctly. It's highly encouraged you make back ups and test in your different staging environments to confirm.

**Comment by uzairleo** [comment]
Got it Also this raised another question that is the GSI setup can be done for one of table in a schemea or not its required for all the available table in schemea since we only need for USER record .

**Comment by uzairleo** [comment]
@Equartey

**Comment by Equartey** [comment]
@uzairleo If I understand the question correctly, you want to know if you must add the `@index` rule to every property on the `User` Model? 

`@index` should only be added to the property(s) you wish to filter or sort by. There is a limit of **20** GSIs you can have on a single table.

**Comment by uzairleo** [comment]
Got it , please double confirm @Equartey it wont effect/impact the already available user query or mutation processes that's already been executing in old build installed with users  and at the same time we change the schema with GSI implementation

**Comment by Equartey** [comment]
@uzairleo This is not a breaking change. GSI enables additional query abilities. 

The schema does not live on the user's device. 

Once you update and push changes to the schema, the GSI will be present in the DynamoDB table. 

DynamoDB handles backfill of the previously created items in the table. So previous entries both persist and function with the new index.

Mutations and queries written before the update will remain unaffected.

**Comment by uzairleo** [comment]
Got it Thanks @Equartey i would suggest to please update your flutter documentation and add these facts regarding usage of queries for frequent purposes specifically the current document is for JS purpose but yeap its closer so we can implement the workaround but it will help users to directly get the information from official docs instead of back and fourth from the different available threads

**Comment by uzairleo** [comment]
Also one thing that we missed during the GSI thread that is why the problem is solved with Custom Query but not with ModelQueries.list() also do you recommend to use the custom Query with this GSI setup for our USER model or not to go with ModelQueries.list() option ?

**Comment by uzairleo** [comment]
@Equartey can you please elaborate the usage of GSI for filtering purpose because in recommended docs which you shared the @index annotation is used just only for the purpose of sorting and it enable to use sortbykey to user. But since our use-case is filtering and i am not able to find any close relation of sorting with filtering our usecase inside docs which lead me to setup GSI in our case by only adding @index to the email in Graphql Schema can you please check the above customQuery which i shared above in issue and let us know what will be the second step for us since you mentioned setting up GSI have two main step 
1-Graphql schema modification
2-Implementation from flutter side but since i am using custom query what will be these changes in our case as compared to sorting guide provided ?

**Comment by uzairleo** [comment]
@Equartey

**Comment by uzairleo** [comment]
UPDATE: @Equartey i setup GSI ( Global seconday index ) but still the issue is there and the same issue is still happening i do test on different environments and everything related to retrieving user email is still the same can you please mention if there is any other fix? 
IF not then we will move to static search from this dynamic search and we will use but i am not sure that feature is supported yet or not that is ----> Getting single record from dynamo db by using email just like we are doing for fetching single object by id

**Comment by Equartey** [comment]
@uzairleo here are two examples of how to perform a query given a specific property (email).

Given this schema: 
```graphql
type USER @model @auth(rules: [{ allow: public }]) {
  id: ID!
  name: String!
  email: String! @index() #GSI
}
```

We can then write a query using the provide model helper functions like so: 
```dart
const userEmail = ""; // define email

// Query for email via Model helper
final request = ModelQueries.list(
USER.classType,
  where: USER.EMAIL.contains(userEmail),
);

final response = await Amplify.API.query(request: request).response;
final users = response.data?.items ?? [];
if (response.errors.isNotEmpty) {
  safePrint('errors: ${response.errors}');
}

safePrint('Query result: $users');
```


Alternatively, since we defined a GSI in our schema, a custom query is generated in AppSync. In this example it would be called `uSERSByEmail`. 

**Note:** The name of this generated function will be determined by GSI definition in the `schema.graphql`. You can verify the generated query name in the AppSync Console. Navigate to AppSync console in your AWS account > select your api > Queries > inspect the explorer for custom queries. 

```dart
const userEmail = ""; // define email

// Custom graphql query
final customRequest = GraphQLRequest<PaginatedResult<USER>>(
  document: r'''
    query QueryByEmail($email: String!) {
      uSERSByEmail(email: $email) {
        items {
          email
          id
          name
        }
      }
    }
    ''',
  variables: {'email': userEmail},
  decodePath: 'uSERSByEmail',
  modelType: const PaginatedModelType(USER.classType),
);

final response = await Amplify.API.query(request: customRequest).response;
final users = response.data?.items ?? [];
if (response.errors.isNotEmpty) {
  safePrint('errors: ${response.errors}');
}

safePrint('Query result: $users');
``` 

[Query Predicate Documentation](https://docs.amplify.aws/lib/graphqlapi/query-data/q/platform/flutter/#query-predicates)

[GSI Custom Query Documenation](https://docs.amplify.aws/cli/graphql/data-modeling/#configure-a-secondary-index)

**Comment by Equartey** [comment]
@uzairleo the name might vary, but when the field is marked as a GSI - `@index()` - a generated query should be visible in the AppSync console. You will still need to setup a graphql request by manually writing the document string, but the above example should allow you to keep your types. 

Let me know if you have any issues.

**Comment by uzairleo** [comment]
I tried this query from amplify dashboard but its not fetching email as the resolver is expecting UserByEmail to be defined somewhere but this query is not defined any where i run this below Query 

```

`uSERSByEmail(email: $email) 
{
 items
 {
email 
id
 name
 } }`
```


Also wanna to mention i do create index on email and set GSI but still the issue is happening and the root cause is userByEmail query is not defined anywhere from backend side while uswrById (id:id)
Is defined and working 


And what i get from this is either we have to override a Resolver or create a lamda  resolver for this simple purpose if i am not wrong ? Can you please confirm

**Comment by Equartey** [comment]
When you made updates to your `schema.graphql`, did you also run `$ amplify codegen models` & `$ amplify push`?

Can you please share your current `schema.graphql`?

**Comment by uzairleo** [comment]
nope i dont need to do codegen model because i am setting up GSI from dynamoDb dashboard directly from table detail
Plus i am running from amplify dashboard content manager section where you guys provide support for testing Queries

**Comment by uzairleo** [comment]
I run the above custom Query and this is the response we are continously getting which clearly show that this custom query is not defined anywhere but the one that is running currently such getUserById(id) is defined in your resolver which is abstracted , here is the response 

```
{
  "data": null,
  "errors": [
    {
      "path": null,
      "locations": [
        {
          "line": 62,
          "column": 3,
          "sourceName": null
        }
      ],
      "message": "Validation error of type FieldUndefined: Field 'UserBYEMAIL' in type 'Query' is undefined @ 'UserBYEMAIL'"
    }
  ]
}
```

Now for this simple use-case such as retrieving single record from DB by non-partitionkey which is fixed through GSI But still not working one have to write a custom LAMDA GRAPHQL resolver for it , i believe @Equartey you have to put enhancement label to this issue to add this functionality as its really basic Query where one will get Item by email non-primarykey by doing GSI setup .

**Comment by uzairleo** [comment]
If you still think there is quick fix available for this please mention here or not then LAMDA RESOLVER will be the only option for us to go with . because i am not able to found any fix for this and the response of the query make it clearer too that each Query we are executing is somewhere defined like listuser , getuserbyid so it cant be any string and should must be defined before someone type this.

**Comment by uzairleo** [comment]
@Equartey ??

**Comment by Equartey** [comment]
@uzairleo Since you setup the GSI manually through DynamoDB, you will need to create the custom query manually. This is why you are unable to find the "UserBYEMAIL" query. 

However, when using the Amplify CLI we handle this for you. The custom GSI query gets generated and deployed for you using the the previously mentioned commands. I verified this when I provided the two examples.

Putting aside the custom query, you can use the first example using query predicates to retrieve the data you want. Here is that example:

```dart 
const userEmail = ""; // define email

// Query for email via Model helper
final request = ModelQueries.list(
USER.classType,
  where: USER.EMAIL.contains(userEmail),
);

final response = await Amplify.API.query(request: request).response;
final users = response.data?.items ?? [];
if (response.errors.isNotEmpty) {
  safePrint('errors: ${response.errors}');
}

safePrint('Query result: $users');
```

**Comment by uzairleo** [comment]
its not working @Equartey i believe i do mention this earlier thats why we are using custom Query with this approach the above you mention the record were available in db but the above method not able to fetch for us . but the custom QUery did that but with some issues so we decided to get a single record by passing email only But look like you miss confused that i do created another issue to make it more clear but you close that and now you are giving method that are out of topic @Equartey it still return list of user and something already we did and results are not Good we need similar method now to avoid all these issues , which lead us to use some similar method just like we have for userbyid 
```
  final request = ModelQueries.get(  USER.classType,
        USERModelIdentifier(id: userId),);
      final response = await Amplify.API.query(request: request).response;


```
This above working for fetching by id but same is not working for email which lead us to use custom Query  because there is no support in fetching single record using dynamodb through email because we try this below its failed with same error i put above 

 ```
final request = ModelQueries.get(  USER.classType,
        USERModelIdentifier(email: email),);
      final response = await Amplify.API.query(request: request).response;
```

**Comment by uzairleo** [comment]
Another thing which you are keep ignoring is why the Query not working in amplify dshboard plus one more thing you mentioned that dont set index from dynamodb i believe if either we set from dynamodb or from codebase both updated there this time i created from codebase and push the code through amplify push but still thats not working . 

Also you are keep ignoring the LAMDA RESOLVER thing ? does there is something to HIDE ?

**Comment by Equartey** [comment]
> ```
> final request = ModelQueries.get(  USER.classType,
>       USERModelIdentifier(email: email),);
> final response = await Amplify.API.query(request: request).response;
> ```

The `ModelQueries.get()` request as written does not work because `.get()` only works on the model's identifier property defined in the `schema.graphql`. In your case `id: ID!`.

In order to use email as the identifier, aka custom primary key, it has to be defined as the the primary key in the schema file. However, I do not recommend this because such a change would be breaking change to your data set. Which you indicated was not a desired outcome.

---

> Also you are keep ignoring the LAMDA RESOLVER thing ? does there is something to HIDE ?

The lambda resolver is not mentioned because there are alternative solutions that are more simple to implement. It can be a solution, but again not recommended.

---

> i believe i do mention this earlier thats why we are using custom Query with this approach the above you mention the record were available in db but the above method not able to fetch for us .

From the previous code sample you were passing a `limit` parameter. This actually sets the amount of items the DynamoDB scans _before_ applying the filter.

Meaning, with out the limit parameter DynamoDB will only be capped by the physical size of the data set (1 MB). Once that limit is reached, you need to paginate the results.

Have you tried removing the `limit` parameter in the original code snippet when testing?

---

> "message": "Validation error of type FieldUndefined: Field 'UserBYEMAIL' in type 'Query' is undefined @ 'UserBYEMAIL'"

This error indicates the @index property was not setup correctly on the model. I'm happy to help troubleshoot this with you.

It would be helpful to see the contents of the `schema.graphql` when you received this error. Can you please provide it here?

**Comment by uzairleo** [comment]
Also you mention `UserByEMAIL` can be any string but according to the available docs and data we collected founded that its not like that in the error its clearly mention that this can't be any string otherwise one will get undefined (that STRING) , and this is something we found by running getUSER(id ) with little modification in the name and thats return the same above error in amplify test-graphql console here it is 
```

{
  getUSERBYuid(id: "USERID"){
    name
    id
    email
    pmtRefNo
  }
}
```
Response 

```
{
  "data": null,
  "errors": [
    {
      "path": null,
      "locations": [
        {
          "line": 62,
          "column": 3,
          "sourceName": null
        }
      ],
      "message": "Validation error of type FieldUndefined: Field 'getUSERBYuid' in type 'Query' is undefined @ 'getUSERBYuid'"
    }
  ]
}
```

But the Query is working for id when the name is `getUser(id:"USERID")` and the query is ```

{
  getTotemUSER(id: "USERID "){
    name
    id
    email
    pmtRefNo
  }
}
```
 So i believe this prove your above concern / point about @index as cause of this response wrong, since index is not required for ID ( primaryKey) but that also not working with different name of query other than getUser(id).

**Comment by Equartey** [comment]
@uzairleo thank you for the verbose context. This helps me understand your situation better.

I'm going to transfer this issue to a team that can better assist you.
